hydra-node/src/Hydra/HeadLogic.hs

```diff
@@ -181,75 +200,96 @@ update Environment{party, signingKey, otherParties} ledger st ev = case (st, ev)
         { contestationPeriod
         , parties = party : otherParties
         }
-  (_, OnChainEvent OnInitTx{contestationPeriod, parties}) ->
+  (ReadyState, OnChainEvent (Observation OnInitTx{contestationPeriod, parties})) ->
     NewState
-      (InitialState (HeadParameters{contestationPeriod, parties}) (Set.fromList parties) mempty)
+      ( InitialState
+          { parameters = HeadParameters{contestationPeriod, parties}
+          , pendingCommits = Set.fromList parties
+          , committed = mempty
+          , previousState = ReadyState
+          }
+      )
       [ClientEffect $ ReadyToCommit $ fromList parties]
   --
-  (InitialState _ remainingParties _, ClientEvent (Commit utxo))
+  (InitialState{pendingCommits}, ClientEvent (Commit utxo))
     | canCommit -> sameState [OnChainEffect (CommitTx party utxo)]
    where
-    canCommit = party `Set.member` remainingParties
-  (InitialState parameters remainingParties committed, OnChainEvent OnCommitTx{party = pt, committed = utxo}) ->
+    canCommit = party `Set.member` pendingCommits
+  (previousState@InitialState{parameters, pendingCommits, committed}, OnChainEvent (Observation OnCommitTx{party = pt, committed = utxo})) ->
     nextState newHeadState $
       [ClientEffect $ Committed pt utxo]
         <> [OnChainEffect $ CollectComTx collectedUTxO | canCollectCom]
    where
-    newHeadState = InitialState parameters remainingParties' newCommitted
-    remainingParties' = Set.delete pt remainingParties
+    newHeadState =
+      InitialState
+        { parameters
+        , pendingCommits = remainingParties
+        , committed = newCommitted
+        , previousState
```

Comment:
All the `previousState@` binds can be avoided by defining `previousState = st`, which we already have in scope.

Reply:
Yes, but we purposely did that because the `st` is a 2-letter identifier which is 500 lines above; so when looking at the branches in isolation, it really comes out of context and, having the rebind is cost-free and immediately clarify what this thing is.